model/sharing/sharing.go

```diff
@@ -1056,206 +1083,423 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
-			if isFirst != isOwner {
-				checks = append(checks, map[string]interface{}{
-					"id":     s.SID,
-					"type":   "invalid_member_status",
-					"member": i,
-					"status": m.Status,
-				})
-			}
+
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_instance_for_member",
+				"member": m.Instance,
+			})
+			return checks
 		}
 
-		if !s.Active {
-			return nil
+		var ms *Sharing
+		if ms, err = FindSharing(member, s.ID()); err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":    s.SID,
+				"type":  "missing_sharing_for_member",
+				"error": err.Error(),
+			})
+			return checks
 		}
 
-		if s.Owner {
-			for i, m := range s.Members {
-				if i == 0 || m.Status != MemberStatusReady {
-					continue
-				}
-				if s.Credentials[i-1].Client == nil {
-					checks = append(checks, map[string]interface{}{
-						"id":     s.SID,
-						"type":   "missing_oauth_client",
-						"member": i,
-						"owner":  true,
-					})
+		if ms.ReadOnly() {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "read_only_member",
+				"member": m.Instance,
+			})
+			return checks
+		}
+
+		memberDocs, err := FindMatchingDocs(member, *rule)
+		if err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_matching_docs_for_member",
+				"member": m.Instance,
+				"error":  err.Error(),
+			})
+			return checks
+		}
+
+		var isMemberDiskFull bool
+		if usage, err := member.VFS().DiskUsage(); err != nil {
+			isMemberDiskFull = usage >= member.VFS().DiskQuota()
+		}
+
+		// Build a map of member docs with their owner's counterpart ids
+		memberKey := s.Credentials[i].XorKey
+		memberDocsByID := make(map[string]couchdb.JSONDoc)
+		for _, doc := range memberDocs {
+			memberDocsByID[XorID(doc.ID(), memberKey)] = doc
+		}
+
+		for _, ownerDoc := range ownerDocs {
+			if memberDoc, found := memberDocsByID[ownerDoc.ID()]; found {
+				if ownerDoc.Rev() != memberDoc.Rev() {
+					if ownerDoc.ShortRev() > memberDoc.ShortRev() && isMemberDiskFull {
+						checks = append(checks, map[string]interface{}{
+							"id":       s.SID,
+							"type":     "disk_quota_exceeded",
+							"instance": m.Instance,
+						})
+						return checks
+					} else if ownerDoc.ShortRev() < memberDoc.ShortRev() && isOwnerDiskFull {
+						checks = append(checks, map[string]interface{}{
+							"id":       s.SID,
+							"type":     "disk_quota_exceeded",
+							"instance": inst.Domain,
+						})
+						return checks
+					} else {
+						checks = append(checks, map[string]interface{}{
+							"id":        s.SID,
+							"type":      "invalid_doc_rev",
+							"member":    m.Instance,
+							"ownerDoc":  ownerDoc,
+							"memberRev": memberDoc.Rev(),
+						})
+					}
+				} else {
+					// It's unnecessary to run these checks if both docs don't
+					// have the same revision in the first place.
```

Comment:
No, I think it would still be useful to check that (at least for `dir_id` and `name`) even if the revision is the same

Reply:
Hum, I'm not sure I get what you mean here. When the revision is the same, we run the other checks. Did you mean we should look at `dir_id` and `name` even if the revisions are **not** the same?

Reply:
My bad, forget that, I didn't read correctly.